test/neural_networks/test_effective_dimension.py

```diff
@@ -0,0 +1,260 @@
+        )
+
+        # set up quantum neural networks
+        num_qubits = 3
+        feature_map = ZFeatureMap(feature_dimension=num_qubits, reps=1)
+        ansatz = RealAmplitudes(num_qubits, reps=1)
+
+        # CircuitQNNs
+        qc = QuantumCircuit(num_qubits)
+        qc.append(feature_map, range(num_qubits))
+        qc.append(ansatz, range(num_qubits))
+
+        def parity(x):
+            return f"{x:b}".count("1") % 2
+
+        circuit_qnn_1 = CircuitQNN(
+            qc,
+            input_params=feature_map.parameters,
+            weight_params=ansatz.parameters,
+            interpret=parity,
+            output_shape=2,
+            sparse=False,
+            quantum_instance=qi_sv,
+        )
+
+        # qnn2 for checking result without parity
+        circuit_qnn_2 = CircuitQNN(
+            qc,
+            input_params=feature_map.parameters,
+            weight_params=ansatz.parameters,
+            sparse=False,
+            quantum_instance=qi_sv,
+        )
+
+        # OpflowQNN
+        observable = PauliSumOp.from_list([("Z" * num_qubits, 1)])
+        opflow_qnn = TwoLayerQNN(
+            num_qubits,
+            feature_map=feature_map,
+            ansatz=ansatz,
+            observable=observable,
+            quantum_instance=qi_sv,
+        )
+
+        self.qnns = {"circuit1": circuit_qnn_1, "circuit2": circuit_qnn_2, "opflow": opflow_qnn}
+
+        # define sample numbers
+        self.n_list = [5000, 8000, 10000, 40000, 60000, 100000, 150000, 200000, 500000, 1000000]
+        self.n = 5000
+
+    @data(
+        # qnn_name, num_inputs, num_params, result
+        ("circuit1", 10, 10, 4.51202148),
+        ("circuit1", 1, 1, 1.39529449),
+        ("circuit1", 10, 1, 3.97371533),
+        ("circuit2", 10, 10, 5.90859124),
+    )
+    @unpack
+    def test_alg_results(self, qnn_name, num_inputs, num_params, result):
+        """Test that the algorithm results match the original code's."""
+        qnn = self.qnns[qnn_name]
+        global_ed = EffectiveDimension(qnn=qnn, weight_samples=num_params, input_samples=num_inputs)
+
+        effdim = global_ed.get_effective_dimension(self.n)
+
+        self.assertAlmostEqual(effdim, result, 5)
+
+    def test_qnn_type(self):
+        """Test that the results are equivalent for opflow and circuit qnn."""
+
+        num_input_samples, num_weight_samples = 1, 1
+        qnn1 = self.qnns["circuit1"]
+        qnn2 = self.qnns["opflow"]
+
+        global_ed1 = EffectiveDimension(
+            qnn=qnn1,
+            weight_samples=num_weight_samples,
+            input_samples=num_input_samples,
+        )
+
+        global_ed2 = EffectiveDimension(
+            qnn=qnn2,
+            weight_samples=num_weight_samples,
+            input_samples=num_input_samples,
+        )
+
+        effdim1 = global_ed1.get_effective_dimension(self.n)
+        effdim2 = global_ed2.get_effective_dimension(self.n)
+
+        self.assertAlmostEqual(effdim1, effdim2, 5)
```

Comment:
This asserts two results are the same - in my mind it would be good to have an expected reference value at least one of these was compared against to ensure that they both don't go bad together and we never notice. I see this is done in other tests too, but I will just comment here.

Reply:
Added an explicit assert.